fix: Pass cursor parameter to server

nbarbettini opened the pull request:
<!-- Provide a brief summary of your changes -->

## Motivation and Context
Fixes https://github.com/modelcontextprotocol/python-sdk/issues/735 by:
- Updating request params to correctly follow the [pagination spec](https://modelcontextprotocol.io/specification/2025-03-26/server/utilities/pagination#request-format)
- Updated tests that assert the payload sent to the server conforms to the spec (regardless of whether the server supports it)

See also: PR comments inline
I'm happy to update the approach if it doesn't fit the style of this repo.

## How Has This Been Tested?
- [x] Updated unit tests
- [x] Manual verification

## Breaking Changes
None - internal changes only, no client interface changes.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

nbarbettini commented on src/mcp/types.py:
```diff
@@ -79,14 +87,6 @@ class Request(BaseModel, Generic[RequestParamsT, MethodT]):
     model_config = ConfigDict(extra="allow")
 
 
-class PaginatedRequest(Request[RequestParamsT, MethodT]):
-    cursor: Cursor | None = None
-    """
-    An opaque token representing the current pagination position.
-    If provided, the server should return results starting after this cursor.
-    """
```
Removed this, since it was only used for the methods in question.

nbarbettini commented on tests/client/conftest.py:
```diff
@@ -0,0 +1,145 @@
+        ]
+
+
+@pytest.fixture
+def stream_spy():
+    """Fixture that provides spies for both client and server write streams.
```
Built this as a fixture so that it can be easily reused in multiple tests.

I needed a way to verify the request payload "on the wire" and this seemed like a good approach, since it did not require any modification of the actual client code (only test fixtures).

ihrpr commented on src/mcp/types.py:
```diff
@@ -53,6 +53,14 @@ class Meta(BaseModel):
     meta: Meta | None = Field(alias="_meta", default=None)
 
 
+class PaginatedRequestParams(RequestParams):
```
Technically with the None check the schema will work, but I think it would be good to follow the spec schema with inheritance: https://github.com/modelcontextprotocol/modelcontextprotocol/blob/main/schema/2025-03-26/schema.ts

```

class PaginatedRequest(
    Request[PaginatedRequestParams | None, MethodT], Generic[MethodT]
):
    """Base class for paginated requests, matching TypeScript PaginatedRequest interface."""

    params: PaginatedRequestParams | None = None
```

ihrpr commented on src/mcp/types.py:
```diff
@@ -359,12 +359,12 @@ class ProgressNotification(
 
 class ListResourcesRequest(
-    PaginatedRequest[RequestParams | None, Literal["resources/list"]]
+    Request[PaginatedRequestParams | None, Literal["resources/list"]]
```
```suggestion
    class ListResourcesRequest(PaginatedRequest[Literal["resources/list"]]):
```

ihrpr commented on src/mcp/types.py:
```diff
@@ -359,12 +359,12 @@ class ProgressNotification(
 
     method: Literal["resources/list"]
-    params: RequestParams | None = None
+    params: PaginatedRequestParams | None = None
```
```suggestion
```

ihrpr commented on src/mcp/types.py:
```diff
@@ -423,12 +423,12 @@ class ListResourcesResult(PaginatedResult):
 
 class ListResourceTemplatesRequest(
-    PaginatedRequest[RequestParams | None, Literal["resources/templates/list"]]
+    Request[PaginatedRequestParams | None, Literal["resources/templates/list"]]
```
same for the rest

ihrpr reviewed (changes requested):
Thank you so much for the fix! 

Would be great to keep types consistent with the [spec](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/main/schema/2025-03-26/schema.ts) 
once that done, ready to merge

nbarbettini commented:
Thanks for the guidance @ihrpr! I've pushed these changes:
- Added a `PaginatedRequest` class to match the spec's inheritance hierarchy
- Updated the `list_xyz()` methods to omit `params` if `cursor is None` (tests still pass)

nbarbettini commented on src/mcp/types.py:
```diff
@@ -358,13 +367,11 @@ class ProgressNotification(
 
     method: Literal["resources/list"]
-    params: RequestParams | None = None
+    params: PaginatedRequestParams | None = None
```
Yes, not enough coffee this morning ☕️
Fixed, thanks!

ihrpr replied:
I can apply changes in a bit and merge if you are busy, sorry for back and forth on this

nbarbettini replied:
Oh, I see what you mean. No problem

ihrpr reviewed (approved):
Thank you soo much!

nbarbettini commented:
No problem, happy to help!
